lib/propolis/src/enlightenment/hyperv/overlay.rs

```diff
@@ -508,6 +677,93 @@ impl OverlayManager {
             &self.acc_mem,
         )
     }
+
+    pub(super) fn export(&self) -> migrate::HyperVOverlaysV1 {
+        self.inner.lock().unwrap().export()
+    }
+
+    pub(super) fn import(
+        self: &Arc<Self>,
+        payload: migrate::HyperVOverlaysV1,
+    ) -> Result<Vec<OverlayPage>, MigrateStateError> {
+        self.inner.lock().unwrap().import(self, payload).map_err(Into::into)
+    }
+}
+
+pub mod migrate {
+    use std::collections::BTreeMap;
+
+    use serde::{Deserialize, Serialize};
+
+    use crate::migrate::{Schema, SchemaId};
+
+    use super::OverlayKind;
+
+    /// A kind of overlay that can appear in a saved overlay manager payload.
+    ///
+    /// NOTE: It is not safe to remove variants from this enum, because older
+    /// Propolis versions may specify them during a migration. Uplevel versions
+    /// that don't support an older overlay kind must decide what to do with
+    /// overlays of that kind (ignore them, translate them, fail migration,
+    /// etc.).
+    #[derive(Debug, Serialize, Deserialize, Ord, PartialOrd, Eq, PartialEq)]
+    #[serde(tag = "kind", content = "value")]
+    pub enum MigratedOverlayKind {
+        Hypercall,
+
+        #[cfg(test)]
+        Test(u8),
+    }
+
+    impl From<OverlayKind> for MigratedOverlayKind {
+        fn from(value: OverlayKind) -> Self {
+            match value {
+                OverlayKind::Hypercall => Self::Hypercall,
+
+                #[cfg(test)]
+                OverlayKind::Test(n) => Self::Test(n),
+            }
+        }
+    }
+
+    impl From<MigratedOverlayKind> for OverlayKind {
+        fn from(value: MigratedOverlayKind) -> Self {
+            match value {
+                MigratedOverlayKind::Hypercall => Self::Hypercall,
+
+                #[cfg(test)]
+                MigratedOverlayKind::Test(n) => Self::Test(n),
+            }
+        }
+    }
+
+    #[derive(Serialize, Deserialize)]
+    pub struct OverlaySetV1 {
+        pub(super) original_contents: Vec<u8>,
+        pub(super) active: MigratedOverlayKind,
+        pub(super) pending: BTreeMap<MigratedOverlayKind, Vec<u8>>,
+    }
+
+    impl std::fmt::Debug for OverlaySetV1 {
+        fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+            f.debug_struct("OverlaySetV1")
+                .field("original_contents", &"<page redacted>")
+                .field("active", &self.active)
+                .field("pending", &self.pending.keys().collect::<Vec<_>>())
+                .finish()
```

Comment:
again, this is where I might like to write
```suggestion
            let Self { original_contents: _, active, pending } = self;
            f.debug_struct("OverlaySetV1")
                .field("original_contents", &"<page redacted>")
                .field("active", active)
                .field("pending", pending.keys().collect::<Vec<_>>())
                .finish()
```

Reply:
Fixed in 2d23e4e.